content/chapters/software-stack/lab/README.md

```diff
@@ -735,6 +735,116 @@ Enter the `support/libcall-syscall/` folder and go through the practice items be
 
 TODO: Quiz
 
+### High-Level Languages
+
+Using the standard C library (libc) frees the programmer from the cumbersome steps of invoking system calls and reimplementing commun features.
+Still, for improved development time and safety, other programming languages can be used, such as Rust, Python, JavaScript.
+Most (if not all) of these high-level programming languages still make use of the standard C library.
+Such that a call to a function in Python would end-up making a call to a function in the standard C library.
+
+The `support/high-level-lang/` folder stores the implementation of a simple "Hello, World!"-printing program in Python.
+We simply invoke the `python` interpreter to run the program:
+
+```
+student@os:~/.../lab/support/high-level-lang$ python hello.py
+Hello, world!
+```
+
+We count the number of functions called from the standard C library and the number of system calls:
+
+```
+student@os:~/.../lab/support/high-level-lang$ ltrace -l 'libc*' python hello.py 2> libc.out
+Hello, world!
+
+student@os:~/.../lab/support/high-level-lang$ wc -l libc.out
+50469 out
+
+student@os:~/.../lab/support/high-level-lang$ strace python hello.py 2> syscall.out
+Hello, world!
+
+student@os:~/.../lab/support/high-level-lang$ wc -l syscall.out
+948 syscall.out
+```
+
+The dynamic standard C library (`libc.so.6`) is a dependency of the Python interpreter (`/usr/bin/python3`):
+
+```
+student@os:~/.../lab/support/high-level-lang$ ldd /usr/bin/python3
+[...]
+        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa6fd6d0000)
+[...]
+```
+
+We can see the complexity of invoking the Python interpreter, resulting in more the 50,000 of library calls being made.
+This means added overhead versus a simple C function.
+However this also means faster development in the Python programming language.
+Each new layer in the software stack simplifies development but adds overhead.
+
+We can use `perf` to compare the running time between the Python and a C "Hello, World!"-printing programs:
+
+```
+student@os:~/.../lab/support/high-level-lang$ sudo perf stat ../static-dynamic/hello
```

Comment:
If we provide students with virtualised environments (VMs / containers), `perf` might not run or run inconsistently. It would be nice to showcase `perf` as well but if virtualisation  hurts us, we can skip it as it only underlines the previous argument.

Reply:
Makes sense. I think `perf` should still run OK on virtualized environments. And it is a tool we could use for other demos as well.

That said, I'm open to alternatives.